### ISSUE TITLE
`I::Item: Clone, J::Item: Clone` triggers FP

### ISSUE THREAD

**Issue by jendrikw** [rust-lang/rust-clippy#8162]
### Summary

This only happens when using both `J::AssociatedType` and one the types is `Self`.

### Lint Name

type_repetition_in_bounds

### Reproducer

I tried this code [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=%23%5Bwarn(clippy%3A%3Atype_repetition_in_bounds)%5D%0A%0Atrait%20T%3A%20Iterator%20%7B%0A%20%20%20%20fn%20f%3CJ%3E(self%2C%20_%3A%20J)%0A%20%20%20%20where%0A%20%20%20%20%20%20%20%20Self%3A%20Sized%2C%0A%20%20%20%20%20%20%20%20Self%3A%3AItem%3A%20Clone%2C%0A%20%20%20%20%20%20%20%20J%3A%20Iterator%2C%0A%20%20%20%20%20%20%20%20J%3A%3AItem%3A%20Clone%2C%0A%20%20%20%20%7B%0A%20%20%20%20%7D%0A%7D):

```rust
#[warn(clippy::type_repetition_in_bounds)]

trait T: Iterator {
    fn f<J>(self, _: J)
    where
        Self: Sized,
        Self::Item: Clone,
        J: Iterator,
        J::Item: Clone,
    {
    }
}
```

I saw this happen:

```
warning: this type has already been used as a bound predicate
 --> src/lib.rs:9:9
  |
9 |         J::Item: Clone,
  |         ^^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/lib.rs:1:8
  |
1 | #[warn(clippy::type_repetition_in_bounds)]
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: consider combining the bounds: `J::Item: Clone + Clone`
```

I expected to see this happen:

No warning.

### Version

```text
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-unknown-linux-gnu
release: 1.57.0
LLVM version: 13.0.0
```

**Comment by jendrikw** [comment]
I've found another example: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=%23!%5Bwarn(clippy%3A%3Atype_repetition_in_bounds)%5D%0A%0Apub%20struct%20ZipAll%3CI%3A%20Iterator%2C%20J%3A%20Iterator%3E(I%2C%20J)%3B%0A%0Aimpl%3CI%3A%20Iterator%2C%20J%3A%20Iterator%3E%20ZipAll%3CI%2C%20J%3E%0Awhere%0A%20%20%20%20I%3A%3AItem%3A%20Clone%2C%0A%20%20%20%20J%3A%3AItem%3A%20Clone%2C%0A%7B%0A%7D%0A)

```rust
#![warn(clippy::type_repetition_in_bounds)]

pub struct ZipAll<I: Iterator, J: Iterator>(I, J);

impl<I: Iterator, J: Iterator> ZipAll<I, J>
where
    I::Item: Clone,
    J::Item: Clone,
{}

```

**Comment by jendrikw** [comment]
I think this is a duplicate (but with a smaller example) of #7360.
